File: server/java/codegen-kotlin/src/test/kotlin/com/metaobjects/generator/kotlin/InheritWithoutRestateGateTest.kt
package com.metaobjects.generator.kotlin

import com.metaobjects.metadata.ktx.loadString
import java.nio.file.Files
import kotlin.test.Test
import kotlin.test.assertTrue

/**
 * GATE / TRIPWIRE for issue #56 — Kotlin codegen must read INHERITED field attributes via
 * the effective accessor, not own-only. This test asserts the CORRECT behavior and is RED
 * until #56 is fixed: a `field.string` that inherits `@maxLength` via `extends`
 * (abstract-field reuse) without restating it must emit `varchar("...", <inherited>)`, but
 * `KotlinTypeMapper` reads `@maxLength` own-only → falls back to the 255 default. See #56.
 */
class InheritWithoutRestateGateTest {

    private val fixture = """{
      "metadata.root": { "package": "test", "children": [
        { "object.entity": { "name": "Base", "abstract": true, "children": [
            { "field.string": { "name": "baseLabel", "@maxLength": 80, "@required": true } }
        ] } },
        { "object.entity": { "name": "Widget", "children": [
            { "field.long":   { "name": "id" } },
            { "field.string": { "name": "label", "extends": "test::Base.baseLabel" } },
            { "source.rdb":   { "@table": "widgets" } },
            { "identity.primary": { "name": "pk", "@fields": ["id"], "@generation": "increment" } }
        ] } }
      ] }
    }""".trimIndent()

    @Test
    fun `inherited maxLength emits the inherited varchar length, not the 255 default`() {
        val outDir = Files.createTempDirectory("ktgate-")
        try {
            val gen = KotlinExposedTableGenerator()
            gen.setArgs(mapOf("outputDir" to outDir.toString()))
            gen.execute(loadString("test", fixture))

            val src = Files.readString(outDir.resolve("test/WidgetTable.kt"))
            // RED until #56 is fixed: inherited @maxLength=80 should win, not the 255 fallback.
            assertTrue("varchar(\"label\", 80)" in src,
                "expected inherited @maxLength=80 on the label column, got:\n$src")
        } finally {
            outDir.toFile().deleteRecursively()
        }
    }
}

File: server/typescript/packages/codegen-ts/test/inherit-without-restate.gate.test.ts
// GATE / TRIPWIRE for issue #56 — codegen must read inherited attributes via the
// EFFECTIVE accessor (attr / attrs), not the OWN-ONLY accessor (ownAttr).
//
// These tests assert the CORRECT behavior and are RED until #56 is fixed. When the
// ownAttr→attr fix lands, they go green and this becomes the permanent regression gate.
// See issue #56 for the full cross-port inventory of own-only reads.

import { describe, test, expect } from "bun:test";
import { MetaDataLoader, InMemoryStringSource } from "@metaobjectsdev/metadata";
import { buildPkMap } from "../src/pk-resolver.js";

async function loadModel(children: unknown[]) {
  const json = JSON.stringify({ "metadata.root": { package: "test", children } });
  const result = await new MetaDataLoader().load([new InMemoryStringSource(json)]);
  expect(result.errors).toEqual([]);
  return result.root;
}

describe("inherit-without-restate gate (#56)", () => {
  // An identity.primary that inherits @fields via node-level `extends` (ADR-0029) without
  // restating it. pk-resolver.ts reads primary.ownAttr(@fields) → undefined → the entity
  // is silently dropped from the PK map → generated schema has NO primary key.
  test("identity.primary extends a base identity (no @fields restated) still yields a PK", async () => {
    const root = await loadModel([
      {
        "object.entity": {
          name: "BaseEntity",
          abstract: true,
          children: [
            { "field.long": { name: "id" } },
            { "identity.primary": { name: "primary", "@fields": ["id"], "@generation": "increment" } },
          ],
        },
      },
      {
        "object.entity": {
          name: "Widget",
          children: [
            { "source.rdb": { "@table": "widgets" } },
            { "field.long": { name: "id" } },
            { "identity.primary": { name: "primary", extends: "test::BaseEntity.primary" } },
          ],
        },
      },
    ]);

    const widget = root.objects().find((o) => o.name === "Widget")!;
    const primary = widget.ownChildren().find((c) => c.type === "identity")!;

    // Root cause documented: effective sees the inherited value; own-only does not.
    expect(primary.attr("fields")).toEqual(["id"]);
    expect(primary.ownAttr("fields")).toBeUndefined();

    // The gate: the entity must have a primary key. RED until #56 is fixed
    // (pk-resolver currently drops Widget because it reads ownAttr(@fields)).
    const pkMap = buildPkMap(root);
    expect([...pkMap.keys()]).toContain("Widget");
  });
});
